### PR TITLE
Suppress template codegen if it is a literal

### PR DESCRIPTION
**Pull request by adamdruppe**
What I was hoping to do is reduce these right out of
the AST entirely, but I couldn't figure out how to do
that right and settled for adding the condition to
skip codegen.

Building Phobos with and without it showed a 10 MB
reduction in the static library. (No change in the
dynamic though, perhaps because the linker did this
same work already.)

In an artificial test, I reduced a 2 MB object file
down to a 200 KB one and the resulting executable down
from 1.7 MB to 740 KB. It also cut link time by a
fraction of a second surely since the linker had less
work to do.

My conditions in the method are trying to be very
conservative to avoid any possible breakage and can
perhaps be loosened up as the implementation matures.

### CODE REVIEW

**Comment by joakim-brannstrom** [comment]
Very interesting. I only "deliver" my applications as statically linked (reasons) and have noticed that they can get quite large. The link time is the most irritating when developing...
If I find the time I want to try your patch to see what the difference is :)

**Comment by adamdruppe** [comment]
Be warned that it might not make a difference at all. I haven't tried it on a real world project yet and my artificial test was specifically designed to trigger it.

```
template foo(int i) {
        private string helper() {
                assert(__ctfe);
                return "lol";
        }

        enum foo = helper();
}

extern(C) void main() {
        string s;
        // generate a bunch of trivial template
        static foreach(i; 0 .. 5000)
                s = foo!i;
}
```

(this isn't exactly the one I talked about in the commit message, which also had some other conditions to help me test edge cases and thus more code inside)

But the result on my linux box

BEFORE:

2.22s, 215,092 RSS build
1.6M a
2.7M a.o

AFTER:

2.04s, 190,996 RSS build
661K a
4.2K a.o


It is my hope that some real world cases will mimic that artificial test because several Phobos templates reduce to the same idea: `isInputRange`, `isIntegral`, `octal`, among others. So perhaps it will help there too.

Note that even if fully successful, this complementary to the `assert(__ctfe)` and `()...` static map stuff Stefan has been talking about lately, which attack different, although related problems.

But I opened the PR at this point to see if it breaks anything in the auto-tester that I overlooked....

**Comment by adamdruppe** [comment]
I just tried swapping compilers and rebuilding an old web app of mine that uses reflection somewhat heavily and there's no change. But... I also specifically avoided trivial templates in that code due to the previous dmd problems.

On the bright side, nothing seems to have broken in it which is a good sign. But it indicates that there may also be zero benefit.

**Review by 12345swordy — CHANGES_REQUESTED** [review]
This needs tests!

**Comment by adamdruppe** [comment]
Tentative result running it on my day job work code: no change in binary size, 3% total build time reduction (though note I was using unoptimized dmd builds both before and after), and a whopping 3 GB less RAM consumption. I wasn't expecting that that ram change since I didn't figure out how to code the memory freeing I wanted! But ran a combined total of seven runs convinced me it is real; apparently it helped anyway.

So far, I've found no regression and while some of my projects show no benefit, this RAM use decrease is alone worth it to me. And it passed the auto-tester as well which is a good sign too that my bypass conditions are sane and it doesn't break user code.

If any of you have some projects and can try before and after, I'd love to know if you see any benefit or any regressions too.

**Review by Geod24 — COMMENTED** [review]
This looks nice. I tested it a tiny bit (based on the diff, not on real world projects), and couldn't find anything wrong. It would be nice if you could add to the documentation was code construct is affected (AFAICS, only templates which alias to `enum`).

File: src/dmd/dtemplate.d [hunk 1]
```diff
@@ -6067,6 +6067,55 @@ extern (C++) class TemplateInstance : ScopeDsymbol
...
 
+    /*
+        The idea of this function is to see if the template instantiation
+        can be 100% replaced with its eponymous member. All other members
+        can be discarded, even in the compiler to free memory, and can be
+        elided entirely from the binary.
+    */
+    final bool isTrivial()
+    {
+        // what if it has a static ctor/dtor?
```
[the remark this answers is not in the document]
Comment (adamdruppe) [reply src/dmd/dtemplate.d:6078]: oh yeah that was the last thing that came to mind while thinking about it and i forgot to delete the comment after adding the loop below.

File: src/dmd/dtemplate.d [hunk 2]
```diff
@@ -6067,6 +6067,55 @@ extern (C++) class TemplateInstance : ScopeDsymbol
         return hash;
     }
 
+    /*
+        The idea of this function is to see if the template instantiation
+        can be 100% replaced with its eponymous member. All other members
+        can be discarded, even in the compiler to free memory, and can be
+        elided entirely from the binary.
```
[the remark this answers is not in the document]
Comment (adamdruppe) [reply src/dmd/dtemplate.d:6074]: I might actually make it private since this is all an internal implementation hack to aid optimization rather than a defined api..... just my later goal might need the function to be called somewhere else (which is why it is a separate function to begin with instead of being inline in the codegen function) and I don't know where that is. might be in another module, idk, if I knew how to do it i would have already done it lol

**Comment by adamdruppe** [comment]
yeah this PR only changes stuff that boils down to `template Foo(args) { enum Foo = ctfe_resolves_to_some_basic_type_or_string; }`. I was extremely conservative to minimize risk of regression and stay inside allowed quality-of-implementation optimizations under the existing D spec. We've had a lot of pain with missing template instantiations in the past causing frustrating linker errors, so I tried to only trigger this in cases where such should be impossible; that the template will NEVER need anything at runtime. (That's why it doesn't even attempt structs and bails out of there's a static ctor. In theory they could be included, but I wasn't confident in my ability to code that correctly and didn't want a missing typeinfo, initializer, or anything else to prop up and ruin someone's day. That's also why this check is BELOW the -allinst check, though it should work just as well with or without that,.)

It is my hope that we can expand the scope in the future once we prove 1) it is viable without breaking anything and 2) it helps build metrics across varied projects while doing no harm to others. (if it doesn't deliver anything anywhere, no point merging, if it harms any users... prolly not worth merging either.)

**Comment by kinke** [comment]
> Building Phobos with and without it showed a 10 MB reduction in the static library.

That sounded too good to be true, so I've quickly tested this with LDC master on Win64.
* druntime-ldc.lib: 4,448,014 bytes (unchanged)
* druntime-ldc-debug.lib: 9,693,212 bytes (unchanged)
* phobos2-ldc.lib: 8,963,026 bytes (-354 bytes)
* phobos2-ldc-debug.lib: 18,448,614 bytes (-1,146 bytes)

**Comment by adamdruppe** [comment]
let me jot down some future direction notes that came to mind while in bed

the restriction on such basic types can obviously be expanded to other built-in types as well (notably arrays of basic types should prolly be an easy one) and in theory, could be any user-defined type that is defined outside this template. The hard part is reliably detecting that.

BUT, even if the implementation sat where it is, there's some interesting things you can do. Imagine getting a UDA. It could be redesigned to do `T getValueUDA(alias fn, R key)(T default_) { return __traits(getAttributes, fn, of type R).tupleof[0] or else default_; }`.

In other words, it searches for a non-basic type, yet returns the basic type member. So the struct is a key and the basic type is the value. Then it'd trigger this existing simple check.... and be a kinda nice api for common cases anyway - this is the way I tend to do it in practice anyway.

Another future direction would be to extend it to aliases as well as enums. If a template returns aliases to its arguments only, then its innards can be cut. But again detecting that was a bit tricky so maybe can come back to it later.

**Comment by adamdruppe** [comment]
On Wed, May 06, 2020 at 05:35:16AM -0700, Martin Kinkelin wrote:
> That sounded too good to be true, so I've quickly tested this with LDC master on Win64.
> * phobos2-ldc-debug.lib: 18,448,614 bytes (-1,146 bytes)

That's like half what I got on my linux box in total size! Mine
went i think from 60 MB down to 50 MB.

I don't remember if this is before or after but like yeah:

-rw-r--r-- 1 me users  67M May  5 17:18 libphobos2.a

But Windows D builds tend to be smaller than Linux ones
anyway, perhaps the librarian is smarter about merging things.
And ldc might already be more intelligent (in separate tests
over the weekend, dmd was giving me (iters*body_size)^2 on static
foreach while ldc, based on same frontend version, was just iters^2.
dmd -o- did that too... so clearly ldc was smarter about avoiding
pointless codegen work and that might be helping it here too.)


Though as long as the patch causes no harm in any other case, even
if it only helps a small fraction, I'm calling that a win.

**Comment by adamdruppe** [comment]
lol a test failed because a llvm site is down.

but i pushed a bit of comment change and changed the protection level to `package` to indicate formally that this is internal and should not be used my dmd as a lib people; it is subject to change at any time.

File: src/dmd/dtemplate.d [hunk 3]
```diff
@@ -6067,6 +6067,70 @@ extern (C++) class TemplateInstance : ScopeDsymbol
...
+        if (v is null)
+            return false;
+
+        if (!(v.storage_class & STC.manifest))
+            return false;
+
+        // I'm only doing basic types here because it is the easiest proof-of-concept
+        // implementation with minimal risk of side effects, but it could likely be
+        // expanded to any type that already exists outside this particular instance.
+        if (!(v.type is Type.tstring || (v.type.isTypeBasic() !is null)))
```
Comment (UplinkCoder) [inline src/dmd/dtemplate.d:6107]: use equals here. they type may not be merged.

**Review by UplinkCoder — APPROVED** [review]
Looks good to me.
I so no reason for not merging it.
let's make sure we mark this as a kludge though.

**Comment by adamdruppe** [comment]
I don't understand why that one test thing failed. It doesn't appear to have any error messages.

**Comment by thewilsonator** [comment]
Yeah the semaphore failure is unrelated.

**Comment by adamdruppe** [comment]
the buildkite failure traces to this in rdmd

>   enforce(!res.output.canFind("compile_force_src"));  // second call will not re-compile

doesn't seem related.... again.

**Comment by adamdruppe** [comment]
and now the azure one fails. says cancellaton.

every time i try to do a PR here the tester seems to play roulette. i don 't understand it.

**Comment by MoonlightSentinel** [comment]
Restarted the tests, the Azure failure got fixed yesterday

**Review by thewilsonator — APPROVED** [review]
Looks OK, could you clean up the comments to be not first person. Also its now green!

File: src/dmd/dtemplate.d [hunk 4]
```diff
@@ -6067,6 +6067,68 @@ extern (C++) class TemplateInstance : ScopeDsymbol
...
+        template foo(args...) {
+            some_basic_type_or_string helper() { .... }
+            enum foo = helper();
+        }
+        ---
+
+        since it was the easiest starting point of implementation but it can and
+        should be expanded more later.
+    */
+    package final bool isTrivial()
```
Comment (WalterBright) [inline src/dmd/dtemplate.d:6092]: I don't think it needs `package`.

File: src/dmd/dtemplate.d [hunk 5]
```diff
@@ -6067,6 +6067,68 @@ extern (C++) class TemplateInstance : ScopeDsymbol
         return hash;
     }
 
+    /**
+        Returns true if the instances' innards are discardable.
```
Comment (WalterBright) [inline src/dmd/dtemplate.d:6071]: Use Ddoc style `Returns:`

File: src/dmd/dtemplate.d (hunk 4 above)
Comment (WalterBright) [reply src/dmd/dtemplate.d:6092]: `isTrivial()` doesn't match the description above. Perhaps `isDiscardable()`?

**Comment by WalterBright** [comment]
Nice insight. It's in a similar vein to my PRs to recognize AliasSeq and Alias templates and short-circuit their evaluation. We need to do more of this.

**Review by WalterBright — CHANGES_REQUESTED** [review]
I'll approve after the two fixes I mentioned.

**Review by WalterBright — APPROVED** [review]

**Comment by adamdruppe** [comment]
On Sat, May 23, 2020 at 02:54:26PM -0700, Walter Bright wrote:
> I don't think it needs `package`.

The reasoning there is just I don't want people using dmd as a library
to depend on this.

But the end goal is to free ast nodes used in the process to determining
the template's final result, just I don't know how to do that without
crashes. Might need access from another module, or maybe not...

I suppose we could make it `private` now and change to `package` later
when that part is figured out though.

File: src/dmd/dtemplate.d (hunk 4 above)
Comment (WalterBright) [reply src/dmd/dtemplate.d]: I'd still rather you removed `package`.

**Comment by adamdruppe** [comment]
On Sat, May 23, 2020 at 02:58:17PM -0700, Walter Bright wrote:
> Nice insight. It's in a similar vein to my PRs to recognize AliasSeq and Alias templates and short-circuit their evaluation. We need to do more of this.

You know, I should study your PR there, that short-circuit code might be
an improvement to this concept as well.

(so what it does now is avoid doing codegen work when it is obviously
unnecessary. but in theory it could do the intermediate template
work in a region/pool allocator then free that whole memory block
in the end, just keeping the final result. no intermediate template
instances, mangles, etc. remaining in memory. even a large recursive
template may be reduced to just one result. it gets trickier when
it reduces to an alias though but I think that's doable too, if the
result is the same as one of its arguments; similar to recognizing
AliasSeq.

I'll have to spend a bunch more time on this and idk when that will
be possible, but it will remain on my todo list.)
